x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsPhysicalOperationProviders.java

```diff
@@ -322,21 +330,53 @@ public PhysicalOperation timeSeriesSourceOperation(TimeSeriesSourceExec ts, Loca
         return PhysicalOperation.fromSource(luceneFactory, layout.build());
     }
 
-    private List<ValuesSourceReaderOperator.FieldInfo> extractFields(
-        List<Attribute> attributes,
-        Function<Attribute, MappedFieldType.FieldExtractPreference> preference
-    ) {
+    List<ValuesSourceReaderOperator.FieldInfo> extractFields(FieldExtractExec fieldExtractExec) {
+        List<Attribute> attributes = fieldExtractExec.attributesToExtract();
         List<ValuesSourceReaderOperator.FieldInfo> fieldInfos = new ArrayList<>(attributes.size());
+        Set<String> nullsFilteredFields = new HashSet<>();
+        fieldExtractExec.forEachDown(EsQueryExec.class, queryExec -> {
+            QueryBuilder q = queryExec.query();
+            if (q != null) {
+                nullsFilteredFields.addAll(nullsFilteredFieldsAfterSourceQuery(q));
+            }
+        });
         for (Attribute attr : attributes) {
             DataType dataType = attr.dataType();
-            var fieldExtractPreference = preference.apply(attr);
+            var fieldExtractPreference = fieldExtractExec.fieldExtractPreference(attr);
             ElementType elementType = PlannerUtils.toElementType(dataType, fieldExtractPreference);
             IntFunction<BlockLoader> loader = s -> getBlockLoaderFor(s, attr, fieldExtractPreference);
-            fieldInfos.add(new ValuesSourceReaderOperator.FieldInfo(getFieldName(attr), elementType, loader));
+            String fieldName = getFieldName(attr);
+            boolean nullsFiltered = nullsFilteredFields.contains(fieldName);
+            fieldInfos.add(new ValuesSourceReaderOperator.FieldInfo(fieldName, elementType, nullsFiltered, loader));
         }
         return fieldInfos;
     }
 
+    /**
+     * Returns the set of fields that are guaranteed to be dense after the source query.
+     */
+    static Set<String> nullsFilteredFieldsAfterSourceQuery(QueryBuilder sourceQuery) {
```

Comment:
Here, we extract fields from the query that can be considered nullsFiltered when reading values. The `nullsFiltered` flag passed to each FieldInfo.

Reply:
It should work for both, but it will be a no-op. Optimizations can be enabled at the codec level, and only possible with the TSDB codec now.